token/services/tokens/tokens.go

```diff
@@ -66,24 +69,45 @@ func (t *Tokens) AppendTransaction(tx Transaction) error {
 		logger.Debugf("transaction [%s on (%s)] is known, extract tokens", txID, tms.ID())
 	}
 	request := tx.Request()
+	if request == nil {
+		logger.Debugf("transaction [%s], no request found, skip it", txID)
+		return nil
+	}
 	if request.Metadata == nil {
 		logger.Debugf("transaction [%s], no metadata found, skip it", txID)
 		return nil
 	}
-	metadata, err := request.GetMetadata()
+
+	ts, err := t.Storage.NewTransaction()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := ts.Rollback(); err != nil {
+			logger.Errorf("failed to rollback [%s]", err)
+		}
+	}()
+	exists, err := ts.TransactionExists(tx.ID())
 	if err != nil {
-		logger.Debugf("transaction [%s], failed to get metadata [%s]", txID, err)
 		return err
 	}
+	if exists {
+		return nil
+	}
 
 	precision := tms.PublicParametersManager().PublicParameters().Precision()
+	md, err := request.GetMetadata()
+	if err != nil {
+		logger.Debugf("transaction [%s], failed to get metadata [%s]", txID, err)
+		return err
+	}
 	if tms.PublicParametersManager().PublicParameters().GraphHiding() {
-		ids := metadata.SpentTokenID()
+		ids := md.SpentTokenID()
 		if logger.IsEnabledFor(zapcore.DebugLevel) {
 			logger.Debugf("transaction [%s] with graph hiding, delete inputs [%v]", txID, ids)
 		}
 		for _, id := range ids {
-			if err := t.TokenStore.DeleteToken(id.TxId, id.Index, txID); err != nil {
+			if err := ts.DeleteToken(id.TxId, id.Index, txID); err != nil {
```

Comment:
I think we could have an optimisation in the query if we had a method `TokenStore.DeleteTokens`.

Reply:
okay